rust/src/detect/mod.rs

```diff
@@ -25,3 +25,24 @@ pub mod stream_size;
 pub mod uint;
 pub mod uri;
 pub mod requires;
+
+use crate::detect::uint::DetectUintData;
+
+/// Enum trait that will be implemented on enums that
+/// derive StringEnum.
+pub trait Enum<T> {
+    /// Return the enum variant of the given numeric value.
+    fn from_u(v: T) -> Option<Self> where Self: Sized;
+
+    /// Convert the enum variant to the numeric value.
+    fn into_u(&self) -> T;
+
+    /// Return the string for logging the enum value.
+    fn to_str(&self) -> &'static str;
+
+    /// Get an enum variant from parsing a string.
+    fn from_str(s: &str) -> Option<Self> where Self: Sized;
+
+    /// Get a detect context for integer keyword.
+    fn to_detect_ctx(s: &str) -> Option<DetectUintData<T>>;
```

Comment:
So you should now be able to do something like:
```
impl<T> TryFrom<&dyn Enum<T>> for DetectUintData<T> {
    type Error = ();

    fn try_from(value: &dyn Enum<T>) -> Result<Self, Self::Error> {
        todo!()
    }
}
```
instead of having this method as part of the trait.

Reply:
Alternatively naming..  This is an enum specifically for use in detection with a direct mapping from a name to an integer value?

Reply:
> So you should now be able to do something like:
> 
> ```
> impl<T> TryFrom<&dyn Enum<T>> for DetectUintData<T> {
>     type Error = ();
> 
>     fn try_from(value: &dyn Enum<T>) -> Result<Self, Self::Error> {
>         todo!()
>     }
> }
> ```
> 
> instead of having this method as part of the trait.

So I think the prototype `fn to_detect_ctx(s: &str) -> Option<DetectUintData<T>>` is right as it takes a string as input, and resorts to generic integer parsing if this is not an enum string 

> This is an enum specifically for use in detection with a direct mapping from a name to an integer value?

The enum is also used in logging (not only for alerts). Was it your question ?

Reply:
It just seems out of place, with respect to naming.  We have a rather generic trait named `Enum`, but then we have this method `to_detect_ctx` that doesn't even take `&self`.

Some more comments in next PR.

Reply:
Maybe the trait can be better named indeed.

> method to_detect_ctx that doesn't even take &self.

But I do not want to use the enumeration : I just want a unique association between integers and strings that goes both ways, and some helper functions around that.
Is derive+trait+enum the best way to do so ? Or do you see another way ?

Reply:
In the gist I posted that has the from traits, maybe a generic helper function?

Reply:
Thans Jason, tried in https://github.com/OISF/suricata/pull/10110/commits/d5f1bf2f1196fad64de37e9ce922aebadc6c6a85